src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/CertificatePal.Windows.PrivateKey.cs

```diff
@@ -62,9 +62,21 @@ public bool HasPrivateKey
 
         public ECDiffieHellman? GetECDiffieHellmanPrivateKey()
         {
+            static ECDiffieHellmanCng? FromCngKey(CngKey cngKey)
+            {
+                if (cngKey.AlgorithmGroup == CngAlgorithmGroup.ECDiffieHellman)
+                {
+                    return new ECDiffieHellmanCng(cngKey, transferOwnership: true);
+                }
+
+                // We might be getting an ECDSA key here. CNG allows ECDH to be either ECDH or ECDSA, however if
+                // the AlgorithmGroup is ECDSA, then it cannot be used for ECDH, even though both of them are ECC keys.
+                return null;
```

Comment:
Null works.  I slightly wonder if an exception would be more appropriate.  Because we have the problem of

* cert.HasPrivateKey: true
* cert.GetECDiffieHellmanPublicKey: non-null (not restricted by KU)
* cert.GetECDiffieHellmanPrivateKey: ???

For everything else, I believe that `(true, non-null)` => non-null (or perhaps an exception)... but I guess it does match the docs "or null if this certificate does not have an ECDiffieHellman private key.".... the private key happens to be ECDSA.

So, yeah.  I'm on the fence and waffling.

Reply:
An exception is what we are doing today (once you get past the assert). The exception is not half-bad, either:

> Keys used with the ECDiffieHellmanCng algorithm must have an algorithm group of ECDiffieHellman.

`null` makes more sense to me since that is what the docs say it will do. CNG has some notion of "Hey your key is ECDSA" and returning null from these methods is what we do for the wrong algorithm. It is only a... little... odd that the public key fetch happens to work because the public half is not constrained.

Reply:
Yeah. We definitely shouldn't crack the key to ask if it's ECDSA before returning the ECDH, since `new X509Certificate2(cert.RawData).GetECDiffieHellmanPublicKey()` wouldn't know to then return null.

So.  Yeah.  OK.  Let's go with null.  ECDH isn't super popular (from a cert) so it... hopefully... doesn't trip anyone up too badly.